### PR TITLE
Implement config versioning

### PR DESCRIPTION
**Pull request by ekatef**
## Changes proposed in this Pull Request

Implements a check to ensure that the current local `config.yaml` is consistent with the version defined in `config.default.yaml`.

## Checklist

- [x] I consent to the release of this PR's code under the AGPLv3 license and non-code contributions under CC0-1.0 and CC-BY-4.0.
- [x] I tested my contribution locally and it seems to work fine.
- [ ] Code and workflow changes are sufficiently documented.
- [ ] Newly introduced dependencies are added to `envs/environment.yaml` and `doc/requirements.txt`.
- [x] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [ ] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by ekatef** [comment]
@davide-f have drafted a possible implementation. Could you please have a look? 

Have added one more field to the version number to track the changes in the config. Not sure it's the most elegant solution, though.

A possible improvement could be to add a check on the differences between the local config and `config.default.yaml`. Sometimes, it feels a good idea to have an opportunity to review the recent updates in the configuration options. However, not sure Snakemake traceback is the best place to place results of such a check. Probably, having an additional helper function would be a better idea. What is your feeling about that?

File: Snakefile [hunk 1]
```diff
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: AGPL-3.0-or-later
 
 import sys
+import yaml
```
Comment (davide-f) [inline Snakefile]: Great draft!

I'd propose to create an auxiliary function into helpers and place there all the code.
The function may by like: check_version(config)
Internally it loads the config.default and then does the magic.
What do you think?

File: config.default.yaml [hunk 2]
```diff
@@ -2,7 +2,7 @@
 #
 # SPDX-License-Identifier: CC0-1.0
 
-version: 0.3.0
+version: 0.3.0.0
```
Comment (davide-f) [inline config.default.yaml]: Please revert :)

File: config.tutorial.yaml [hunk 3]
```diff
@@ -2,7 +2,7 @@
 #
 # SPDX-License-Identifier: CC0-1.0
 
-version: 0.3.0
+version: 0.3.0.0
```
Comment (davide-f) [inline config.tutorial.yaml]: Please revert

**Review by davide-f — COMMENTED** [review]
Great @ekatef :D
Beyond the minor comments, please add a release doc :)

File: config.default.yaml (hunk 2 above)
Comment (ekatef) [reply config.default.yaml]: Done!

File: config.tutorial.yaml (hunk 3 above)
Comment (ekatef) [reply config.tutorial.yaml]: Done :)

File: Snakefile (hunk 1 above)
Comment (ekatef) [reply Snakefile]: Agree, revised.

**Comment by ekatef** [comment]
> Great @ekatef :D Beyond the minor comments, please add a release doc :)

Done 🙂 Thanks a lot for the review!

I'm thinking what would be the best way to remind that a release is needed every time, when the config structure gets changed. Adding a point for that to PR template sounds like not a really good idea, but I don't have better ones at the moment. What is your perspective on that?

File: Snakefile (hunk 1 above)
Comment (davide-f) [inline Snakefile]: Can we drop this line? it seems is not needed

File: scripts/_helpers.py [hunk 4]
```diff
@@ -27,6 +27,28 @@
...
+def check_config_version(config):
+    """
+    Check that a version of the local config.yaml matches to the actual config
+    version as defined in config.default.yaml.
+    """
+
+    # using snakemake capabilities to deal with yanl configs
+    with open("config.default.yaml", "r") as f:
+        actual_config = yaml.safe_load(f)
+    actual_config_version = actual_config.get("version")
```
Comment (davide-f) [inline scripts/_helpers.py]: I'd propose to have the file path as parameter here, like:

```suggestion
def check_config_version(config, fp_config="config.default.yaml"):
    """
    Check that a version of the local config.yaml matches to the actual config
    version as defined in config.default.yaml.
    """

    # using snakemake capabilities to deal with yanl configs
    with open(fp_config, "r") as f:
        actual_config = yaml.safe_load(f)
    actual_config_version = actual_config.get("version")
```

**Review by davide-f — COMMENTED** [review]
Great @ekatef :D

I've added a minor comment. Moreover there is a minor merge conflict to fix.
This PR is ready to go afterwards. If you like to include these minor changes, that would be nice but the functionality is there.
Afterwards, if CI passes, feel free to merge yourself :)

File: Snakefile (hunk 1 above)
Comment (ekatef) [reply Snakefile]: True, have forgotten to remove it. Thanks!

File: scripts/_helpers.py (hunk 4 above)
Comment (ekatef) [reply scripts/_helpers.py]: Agree, done!

**Comment by ekatef** [comment]
> Great @ekatef :D
> 
> I've added a minor comment. Moreover there is a minor merge conflict to fix. This PR is ready to go afterwards. If you like to include these minor changes, that would be nice but the functionality is there. Afterwards, if CI passes, feel free to merge yourself :)

@davide-f fixed! Thank you 😄 Ok, then I'm merging it.

As a comment, the current implementation writes an error in logs in case the config is outdated, but don't terminate execution. Leaving that for now, while it can be improved further when introducing better management of the configs.

**Pull request merged**